File: backend/app/schemas/crime.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import date, time, datetime
from enum import Enum

class CrimeType(str, Enum):
    theft = "theft"
    assault = "assault"
    # ... add all crime types

class CrimeStatus(str, Enum):
    pending = "pending"
    assigned = "assigned"
    # ... add all statuses

class Priority(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"
    critical = "critical"

class CrimeBase(BaseModel):
    title: str
    description: str
    crime_type: CrimeType
    location: str
    district: str
    latitude: float
    longitude: float
    incident_date: date
    incident_time: time

class CrimeCreate(CrimeBase):
    pass

class CrimeUpdate(BaseModel):
    status: Optional[CrimeStatus] = None
    priority: Optional[Priority] = None
    investigation_notes: Optional[str] = None

class Crime(CrimeBase):
    id: int
    crime_number: str
    status: CrimeStatus
    priority: Priority
    created_at: datetime

    class Config:
        orm_mode = True

File: frontend/src/types/crime.ts
export const crimeTypes = ["theft", "assault", "robbery", "fraud", "cyber_crime", "drug_offense", "murder", "kidnapping", "domestic_violence", "traffic_violation"] as const;
export type CrimeType = typeof crimeTypes[number];

export const crimeStatuses = ["pending", "assigned", "under_investigation", "evidence_collection", "awaiting_approval", "resolved", "rejected"] as const;
export type CrimeStatus = typeof crimeStatuses[number];

export const priorities = ["low", "medium", "high", "critical"] as const;
export type Priority = typeof priorities[number];

export interface Evidence {
    id: number;
    original_filename: string;
    content_type: string;
    file_size: number;
    uploaded_at: string;
    download_url: string;
}

export interface Crime {
    id: number;
    crime_number: string;
    title: string;
    description: string;
    crime_type: CrimeType;
    status: CrimeStatus;
    priority: Priority;
    location: string;
    district: string;
    state: string;
    latitude: number;
    longitude: number;
    incident_date: string;
    incident_time: string;
    created_at: string;
    updated_at: string;
    citizen_id: number;
    assigned_officer_id: number | null;
    assigned_at: string | null;
    evidence_items?: Evidence[];
}

export type CrimeFormValues = Omit<Crime, "id" | "crime_number" | "status" | "priority" | "created_at" | "updated_at" | "citizen_id" | "assigned_officer_id" | "assigned_at" | "evidence_items">;

export type CrimeUpdatePayload = Partial<Pick<Crime, "status" | "priority">>;

export type CrimeFilters = Partial<Pick<Crime, "crime_type" | "status" | "priority" | "district" | "incident_date"> & { query?: string }>;

export interface DashboardStats {
    [key: string]: number;
}

export const labelize = (s: string) => s.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase());